### PR TITLE
test(manager): implement test for no-delta backup

### PR DESCRIPTION
**Pull request by mikliapko**
Closes https://github.com/scylladb/scylla-manager/issues/3936

Changes:
- The test that runs two backups one by one, no inserting any data to DB between them, and checks the 2nd backup takes time -> 0 has been added;
- Corresponding config file for 100GB dataset, test-case yaml and .jenkinsfile have been added as well.

### Testing

- [x] https://jenkins.scylladb.com/view/scylla-manager/job/manager-master/job/no-delta-backup-test/8/

### PR pre-checks (self review)

- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

### CODE REVIEW

File: mgmt_cli_test.py
Comment (soyacz) [inline mgmt_cli_test.py:622]: I think you might also `flush` memtables to disk before doing backup.

Comment (soyacz) [inline mgmt_cli_test.py:622]: I wonder if disable_autocompaction disables also tablets load balancer and tablet split/merge mechanism.

File: mgmt_cli_test.py
Comment (soyacz) [inline mgmt_cli_test.py:645]: maybe worth to delete backup 1 to see if it doesn't break backup 2?

File: configurations/manager/10GB_dataset.yaml
Comment (soyacz) [inline configurations/manager/10GB_dataset.yaml:6]: this overrides db type from `test-cases/manager/manager-no-delta-backup-set-dataset.yaml` which one is valid?

File: mgmt_cli_test.py
Comment (Michal-Leszczynski) [inline mgmt_cli_test.py:622]: I'm not sure if disabling autocompaction stops ongoing compactions, or does it just prevent new ones from starting. If it does not stop them, then it's possible that there would be a compaction running during the backup. It would result in creating new files that needs to be uploaded as a part of the second backup.

You could create table, disable autocompaction and only after that run the c-s command to fill the table.

Comment (Michal-Leszczynski) [reply mgmt_cli_test.py:622]: I believe that taking a snapshot flushes table on its own.

Comment (Michal-Leszczynski) [reply mgmt_cli_test.py:622]: That's an interesting question. @denesb do you know the answer?

File: mgmt_cli_test.py
Comment (mikliapko) [reply mgmt_cli_test.py:645]: IMO It might be a good improvement for the test.
@Michal-Leszczynski What do you think about it?

File: mgmt_cli_test.py
Comment (mikliapko) [reply mgmt_cli_test.py:622]: I suppose we are pretty save in this regard. 
Here is the output of `sctool progress ...` for the two consecutive backup operations:
```
Running command "sudo sctool  -c 75f34691-49ca-4fdb-bfa3-d56229b4e809 progress backup/2a9c1120-6074-474b-80e5-001393f359f7" finished with status 0
Run:		96414835-5676-11ef-ac67-02a275dbf199
Status:		DONE
Start time:	09 Aug 24 17:41:19 UTC
End time:	09 Aug 24 17:43:39 UTC
Duration:	2m19s
Progress:	100%
Snapshot Tag:	sm_20240809174120UTC
Datacenters:	
- us-east

╭─────────────┬──────────┬─────────┬─────────┬──────────────┬────────╮
│ Host        │ Progress │    Size │ Success │ Deduplicated │ Failed │
├─────────────┼──────────┼─────────┼─────────┼──────────────┼────────┤
│ 10.12.0.64  │     100% │ 11.406G │ 11.406G │            0 │      0 │
│ 10.12.0.69  │     100% │ 11.406G │ 11.406G │            0 │      0 │
│ 10.12.2.186 │     100% │ 11.406G │ 11.406G │            0 │      0 │
╰─────────────┴──────────┴─────────┴─────────┴──────────────┴────────╯
```
No-delta backup operation (the test was executed from the PR branch):
```
Running command "sudo sctool  -c 75f34691-49ca-4fdb-bfa3-d56229b4e809 progress backup/783ae129-6762-46cb-98bb-4321aaf1d1e6"...

Run:		1045364a-5677-11ef-ac68-02a275dbf199
Status:		DONE
Start time:	09 Aug 24 17:44:44 UTC
End time:	09 Aug 24 17:44:47 UTC
Duration:	2s
Progress:	100%
Snapshot Tag:	sm_20240809174444UTC
Datacenters:	
- us-east

╭─────────────┬──────────┬─────────┬─────────┬──────────────┬────────╮
│ Host        │ Progress │    Size │ Success │ Deduplicated │ Failed │
├─────────────┼──────────┼─────────┼─────────┼──────────────┼────────┤
│ 10.12.0.64  │     100% │ 11.406G │ 11.406G │      11.406G │      0 │
│ 10.12.0.69  │     100% │ 11.406G │ 11.406G │      11.406G │      0 │
│ 10.12.2.186 │     100% │ 11.406G │ 11.406G │      11.406G │      0 │
╰─────────────┴──────────┴─────────┴─────────┴──────────────┴────────╯
```
As you can see, all the data is deduplicated in case of second backup and no size changes happened.

File: configurations/manager/10GB_dataset.yaml
Comment (mikliapko) [reply configurations/manager/10GB_dataset.yaml]: Fixed, removed `instance_type_db` from test-cases yaml

File: mgmt_cli_test.py
Comment (soyacz) [reply mgmt_cli_test.py:622]: If this is a problem, we have something implemented in perf tests `_wait_no_compactions_running` - could be moved and reused if necessary.

File: configurations/manager/10GB_dataset.yaml
Comment (mikliapko) [reply configurations/manager/10GB_dataset.yaml]: Ah, okay, test-cases yaml should contain instance_type_db parameter, swapped

File: mgmt_cli_test.py
Comment (denesb) [reply mgmt_cli_test.py:622]: No, disable_autocompaction only disables compactions (and waits for any running one to finish).

Comment (denesb) [reply mgmt_cli_test.py:622]: Note that this only refers to automatic compaction -- user can still start compaction via API (e.g. major, scrub, etc).

File: mgmt_cli_test.py
Comment (Michal-Leszczynski) [reply mgmt_cli_test.py:645]: I don't fully understand what "delete backup 1" means.
Is it about deleting the task? Or is it about deleting the snapshot?
We could do the first one, although this wouldn't be relevant to the test.
We can't do the second one, as then the backup would need to be fully re-uploaded.

Comment (mikliapko) [reply mgmt_cli_test.py:645]: If we can't do the second one, then it actually doesn't make sense.

Comment (soyacz) [reply mgmt_cli_test.py:645]: ok, so it means delta-backup strongly relies on previous snapshots and these can't be deleted. I hope it's documented somewhere.

Comment (mikliapko) [reply mgmt_cli_test.py:645]: Hm, just curious, in `backup delete` [doc](https://manager.docs.scylladb.com/stable/sctool/backup.html#backup-delete) there is such info:
```
Deduplicated files are persisted unless their reference count drops to zero.
```
I suppose the flow with running backup deletion with `sctool backup delete --snapshot-tag BACKUP_1_TAG` should keep deduplicated files required for backup 2 and we are kinda safe to restore from it.
@Michal-Leszczynski What do you think?

Comment (mikliapko) [reply mgmt_cli_test.py:645]: Whatever, guys @soyacz @Michal-Leszczynski, if you don't mind I'd suggest to move this investigation and possible test improvement to a separate ticket to not block this PR now since we don't have much time at the moment to work on it?

**Review by Michal-Leszczynski — APPROVED** [review]

File: configurations/manager/10GB_dataset.yaml
Comment (fruch) [reply configurations/manager/10GB_dataset.yaml]: just to clarify, anything under `test-cases` folder is checked, and expected to be full functioning cases as it, hence all the mandatory field are checked

anything under `configurations` is known to be partial configuration, and not being tested by CI for correctness and missing options.

File: mgmt_cli_test.py
Comment (fruch) [reply mgmt_cli_test.py:622]: in dtest when disabling compactions, we wait for compactions to finish before moving on.

even that it's not completely clear if backup can work as expected during compactions, why manager is stop the compactions on it's own on the tables is currently working on ?, why does the user need to care about it ?

File: configurations/manager/10GB_dataset.yaml
Comment (soyacz) [inline configurations/manager/10GB_dataset.yaml:1]: If there are no plans to reuse this 10GB dataset in any other test with these settings, better move it to `test-cases/manager/manager-no-delta-backup-set-dataset.yaml`

**Review by soyacz — APPROVED** [review]
LGTM

File: configurations/manager/10GB_dataset.yaml
Comment (soyacz) [reply configurations/manager/10GB_dataset.yaml:1]: I mean whole file, not only `round_robin` setting.

Comment (mikliapko) [reply configurations/manager/10GB_dataset.yaml:1]: I'd prefer to keep it here in separate file.
It allows us to be more flexible and reuse this dataset in another tests or run this test with different dataset.
It's sometimes a case for the runs which we trigger manually from time to time.

File: mgmt_cli_test.py
Comment (mikliapko) [reply mgmt_cli_test.py:622]: > in dtest when disabling compactions, we wait for compactions to finish before moving on.

I'll add this check in the test to wait compaction to finish.

> even that it's not completely clear if backup can work as expected during compactions, why manager is stop the compactions on it's own on the tables is currently working on ?, why does the user need to care about it ?

The test checks the specific case of data deduplication for consecutive no-delta backup. This scenario might be achieved only with compaction disabled. See more details in the [comment](https://github.com/scylladb/scylla-manager/issues/3936#issuecomment-2277611709).

cc @karol-kokoszka

File: configurations/manager/10GB_dataset.yaml
Comment (mikliapko) [reply configurations/manager/10GB_dataset.yaml:1]: okay, I got the point, the test should be fully functioning with only test-cases configuration provided, will move

Comment (mikliapko) [reply configurations/manager/10GB_dataset.yaml]: Done

**Comment by mikliapko** [comment]
Comments have been addresses, retested here:
https://jenkins.scylladb.com/view/scylla-manager/job/manager-master/job/no-delta-backup-test/8/

**Review by karol-kokoszka — APPROVED** [review]

**Comment by roydahan** [comment]
> @soyacz @fruch I suppose every comment is addressed. Could you please take a look once again?

Both are on PTO, coming back next week if it can wait.

**Comment by mikliapko** [comment]
> Both are on PTO, coming back next week if it can wait.

yes, it can wait, no problem

**Review by soyacz — APPROVED** [review]
LGTM

**Comment by soyacz** [comment]
need one more review from e.g. @grzywin to be able to merge this one.

File: mgmt_cli_test.py
Comment (grzywin) [reply mgmt_cli_test.py:635]: @mikliapko

Comment (mikliapko) [reply mgmt_cli_test.py:635]: Yes, I agree that it makes code less redundant, but I don't think it really adds to readability.
Moreover, from my point of view, it adds to complexity making it harder to understand the test whose main idea is to run two backups one by one. I want to keep them as two separate steps to emphasize that.
One more thing, in the further [steps](https://github.com/mikliapko/scylla-cluster-tests/blob/incremental-backup/mgmt_cli_test.py#L642:L646) the variable `backup_task_2` is used several times. Again, for easier test understanding it's better to have it as `backup_task_2` not simply `backup_task`.

@grzywin What do you think?

Comment (grzywin) [reply mgmt_cli_test.py:635]: @mikliapko Personally, I prefer to loop such redundant code fragments. It is easier for me to read and understand, but it is not a 'must'. I missed the fact that you are using `backup_test_2` several times later on, in such case I agree to leave it as it is.

**Review by grzywin — APPROVED** [review]
lgtm

**Comment by mikliapko** [comment]
> need one more review from e.g. @grzywin to be able to merge this one.

@soyacz I think we are good to merge now, thanks!

**Comment by soyacz** [comment]
hmm, still can't - @fruch can you help out merging this one? Need 2 approves from reviewers with write access

**Review by fruch — APPROVED** [review]

**Pull request merged**
